### PR TITLE
feat(whatsapp): render Baileys rich messages (template, interactive, list, poll, location, contact)

### LINKED ISSUE #274: Bridge do Baileys marca mensagens interactive/template/list/buttons como is_unsupported e perde o texto + URL de CTA
## Resumo

Mensagens recebidas de WhatsApp dos tipos `interactiveMessage`, `templateMessage`, `buttonsMessage`, `listMessage` (e vários outros que a biblioteca `@whiskeysockets/baileys` já decodifica) chegam ao Chatwoot pelo bridge do Baileys com `content: null` e `content_attributes.is_unsupported: true`. O usuário vê apenas o fallback "Esta mensagem não é suportada. Você pode ver esta mensagem no aplicativo", mesmo que o Baileys tenha decodificado corretamente o corpo e o link/CTA da mensagem.

Isso é especialmente problemático para **notificações transacionais enviadas por remetentes na WABA (Cloud API / Business API)** — por exemplo, templates utility com botão de URL ("fatura disponível", "pedido confirmado", cobranças, etc.). O inbox do Chatwoot roda Baileys (não-oficial), mas o remetente roda Cloud API, então o bridge do lado receptor descarta em silêncio conteúdo útil que o Baileys **já tinha decodificado**.

## Caso real observado

- Inbox `Channel::Whatsapp`, `provider: baileys`, conectado a um número pessoal da operação.
- Uma seguradora enviou uma notificação de fatura via WABA com estrutura tipo:
  ```
  Prezados,

  Segue fatura:
  Seguradora: TOO SEGUROS
  Vencimento: 30/04/2026
  ...
  Arquivos disponíveis:
  fatura-too-03-2026-xxxxx.pdf, boleto-too-03-2026-xxxxx.pdf
  ...
  Link para download: https://segurosja.com.br/.../faturamento.php?token_confirma=...
  ```
- No Chatwoot a mensagem foi armazenada como:
  ```json
  {
    "content": null,
    "content_type": "text",
    "content_attributes": {
      "is_unsupported": true,
      "external_created_at": 1776885083
    },
    "source_id": "<waid>"
  }
  ```
- A UI renderizou só o balão de "mensagem não suportada". O agente precisou abrir o WhatsApp no celular da operação para ler a fatura — perdendo o Chatwoot como system of record.

## Reprodução

1. Configurar um inbox `Channel::Whatsapp` com `provider: baileys`.
2. Enviar para esse número, a partir de outro número que usa **WABA Cloud API / Business API**, uma das opções:
   - Template utility com URL button (`templateMessage` / `hydratedTemplateMessage`).
   - Mensagem interactive com CTA URL button (`interactiveMessage` com `nativeFlowMessage` / `cta_url`).
   - List message (`listMessage`).
   - Buttons message (`buttonsMessage`).
3. Observar a mensagem salva — `content` é `null` e `is_unsupported` é `true`.

## Causa raiz

O método `message_type` em [`app/services/whatsapp/baileys_handlers/helpers.rb`](https://github.com/fazer-ai/chatwoot/blob/main/app/services/whatsapp/baileys_handlers/helpers.rb) mantém uma whitelist fechada:

- `conversation`
- `extendedTextMessage`
- `imageMessage`
- `audioMessage`
- `videoMessage`
- `documentMessage` / `documentWithCaptionMessage`
- `stickerMessage`
- `reactionMessage`
- `editedMessage`
- `contactMessage` (apenas se vcard tem `waid=`)
- `protocolMessage`
- `messageContextInfo`

Qualquer outro tipo cai no `else` final e retorna `'unsupported'`. Em [`app/services/whatsapp/baileys_handlers/concerns/message_creation_handler.rb`](https://github.com/fazer-ai/chatwoot/blob/main/app/services/whatsapp/baileys_handlers/concerns/message_creation_handler.rb), esse `'unsupported'` aciona `content_attributes[:is_unsupported] = true` com `content = nil`.

Tipos ausentes da whitelist, porém decodificáveis pela lib Baileys e comumente entregues via WABA:

- `interactiveMessage` / `interactiveResponseMessage`
- `buttonsMessage` / `buttonsResponseMessage`
- `listMessage` / `listResponseMessage`
- `templateMessage` / `templateButtonReplyMessage` / `hydratedTemplateMessage`
- `pollCreationMessage` / `pollUpdateMessage`
- `productMessage` / `orderMessage`

## Ajuste proposto

1. Estender `message_type` (e criar um extractor complementar) em `helpers.rb` para detectar `interactiveMessage`, `buttonsMessage`, `listMessage`, `templateMessage` (+ `hydratedTemplateMessage`) e seus `*ResponseMessage`.
2. Extrair e concatenar numa única string:
   - Corpo (`body.text` / `hydratedContentText` / `contentText`).
   - Footer (`footer.text` / `hydratedFooterText`).
   - Itens da lista (título + descrição de cada row).
   - Botões (label + URL do CTA / payload dos quick-reply).
3. Retornar `'text'` para que o pipeline grave o `content` normalmente, em vez de marcar como unsupported.

Esboço (pseudocódigo):

```ruby
# em helpers.rb
def message_type
  msg = unwrap_ephemeral_message(@raw_message[:message])
  return 'text' if msg.key?(:conversation) || msg.dig(:extendedTextMessage, :text).present?
  return 'image'    if msg.key?(:imageMessage)
  return 'audio'    if msg.key?(:audioMessage)
  return 'video'    if msg.key?(:videoMessage)
  return 'file'     if msg.key?(:documentMessage) || msg.key?(:documentWithCaptionMessage)
  return 'sticker'  if msg.key?(:stickerMessage)
  return 'reaction' if msg.key?(:reactionMessage)
  return 'edited'   if msg.key?(:editedMessage)
  return 'protocol' if msg.key?(:protocolMessage)
  return 'context'  if msg.key?(:messageContextInfo) && msg.keys.count == 1

  # novo: tipos interativos com texto recuperável
  return 'text' if interactive_like?(msg)

  'unsupported'
end

def interactive_like?(msg)
  msg.key?(:interactiveMessage) ||
    msg.key?(:interactiveResponseMessage) ||
    msg.key?(:buttonsMessage) ||
    msg.key?(:buttonsResponseMessage) ||
    msg.key?(:listMessage) ||
    msg.key?(:listResponseMessage) ||
    msg.key?(:templateMessage) ||
    msg.key?(:templateButtonReplyMessage)
end

def extract_interactive_text(msg)
  if (im = msg[:interactiveMessage])
    [im.dig(:header, :title), im.dig(:body, :text), im.dig(:footer, :text),
     im.dig(:nativeFlowMessage, :buttons)&.map { |b| b.dig(:buttonParamsJson) }&.join("
")].compact.join("
")
  elsif (tm = msg[:templateMessage])
    ht = tm[:hydratedTemplate] || tm.dig(:fourRowTemplate)
    [ht&.dig(:hydratedTitleText), ht&.dig(:hydratedContentText), ht&.dig(:hydratedFooterText),
     ht&.dig(:hydratedButtons)&.map { |b| b.dig(:urlButton, :url) || b.dig(:quickReplyButton, :displayText) }&.join("
")].compact.join("
")
  elsif (bm = msg[:buttonsMessage])
    [bm.dig(:contentText), bm.dig(:footerText), bm[:buttons]&.map { |b| b.dig(:buttonText, :displayText) }&.join(" | ")].compact.join("
")
  elsif (lm = msg[:listMessage])
    rows = lm[:sections]&.flat_map { |s| s[:rows]&.map { |r| "• #{r[:title]} — #{r[:description]}" } }&.join("
")
    [lm[:title], lm[:description], lm[:footerText], rows, lm[:buttonText]].compact.join("
")
  end
end
```

O chamador em `message_creation_handler.rb` precisa então pegar esse texto extraído (ex.: `content = helpers.extract_interactive_text(raw_msg)` quando o tipo for interactive-like) antes de cair na criação da mensagem.

## Follow-ups opcionais

- `pollCreationMessage` → gerar nota privada com as opções + correlacionar `pollUpdateMessage` de volta.
- `productMessage` / `orderMessage` → preservar como texto com nome do produto, quantidade e preço.

## Ambiente

- `fazer-ai/chatwoot` `main` (self-hosted).
- Provider `baileys` em `Channel::Whatsapp`.

## Impacto

Sempre que um remetente não-Baileys (qualquer um na Cloud API / Business API oficial) manda mensagem para um inbox Baileys, conteúdo operacional crítico (faturas, confirmações de pedido, prompts de 2 etapas, cobranças, lembretes com link) é descartado. Agentes precisam abrir o WhatsApp no celular para ler — o que derrota o Chatwoot como fonte única de histórico de atendimento.

### PR DESCRIPTION
**Pull request by gabrieljablonski**
WhatsApp messages that carry structured content (marketing/utility templates, interactive button menus, list menus, polls, shared locations and shared contacts) used to arrive through the Baileys bridge as an empty "unsupported" bubble, even though Baileys had already decoded the body, buttons and links. Agents now see the full message: the text, a structured card with clickable buttons (links open in a new tab, phone numbers dial), any media that came in the header (e.g. an invoice PDF or banner image), and native bubbles for locations (map) and shared contacts.

This builds directly on @otimizzeai's template work in #313, which is preserved in the history and credited via co-authorship.

## Closes

- Closes #274
- Builds on / supersedes #313 (credit to @otimizzeai)

## How to test

1. Point a Baileys-backed WhatsApp inbox at a number that receives structured messages.
2. From a WABA/Cloud-API sender, send to that inbox:
   - a utility/marketing **template** with body + footer + a URL button and a call button;
   - a template with **media in the header** (image or an invoice PDF);
   - an **interactive** message (button menu / app-download CTA / list menu);
   - a **poll**;
   - a shared **location** and one or more shared **contacts**.
3. In the dashboard, confirm each renders as a readable card (no "unsupported" badge): the URL button opens in a new tab, the call button dials, header media shows inline, the location opens the map bubble, and each shared contact shows in its own contact bubble.

> Header media for templates also requires the matching `baileys-api` bridge change (separate PR) so the bridge extracts and exposes the header attachment.

## What changed

- New `Whatsapp::Baileys::RichMessageParser` PORO normalizes every rich shape (template / interactive nativeFlow / buttons / list / their response variants / poll) into `{ type, title, body, footer, buttons:[{text,url,phone}] }`, plus a flat text rendering for `content`/previews. Defensive JSON parsing means an unexpected payload degrades to `is_unsupported` (the old behavior) and is logged, instead of raising.
- Handles the real WABA "interactive template" shape nested under `templateMessage.interactiveMessageTemplate`.
- Surfaces media in a rich header as a native attachment (image/video/file).
- Locations and shared contacts are stored as native `location`/`contact` attachments (one message per contact), mirroring the Cloud provider.
- New `RichMessage.vue` bubble renders the card; URLs are http(s)-validated like `ReferralCard`.
- CTWA `externalAdReply` carried on rich messages flows into `referral` via the existing helper, so first-touch attribution keeps working.

- - -
This change is [Reviewable](https://reviewable.io/reviews/fazer-ai/chatwoot/315)

## Summary by CodeRabbit

* **New Features**
  * Rich message rendering: card-style messages with title, formatted HTML body, attachment chips, and actionable buttons/links (safe URL handling).
  * Contact messages: native contact cards created per contact.
  * Location messages: native location attachments with coordinates and place info.

* **Tests**
  * Added comprehensive specs covering rich parsing, rendering, contact and location message handling, and edge cases (media headers, malformed params, reply anchoring).

### CODE REVIEW

**Comment by gabrieljablonski** [comment]
@coderabbitai review

File: app/javascript/dashboard/components-next/message/bubbles/RichMessage.vue [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+          "
+        >
+          <Icon
+            v-if="button.icon"
+            :icon="button.icon"
+            class="size-3 shrink-0"
+          />
+          <span>{{ button.text }}</span>
+        </component>
+      </div>
```
[the remark this answers is not in the document]
Comment (gabrieljablonski) [reply app/javascript/dashboard/components-next/message/bubbles/RichMessage.vue:96]: Fixed — rel is now bound conditionally (`:rel="button.isLink ? 'noopener noreferrer' : undefined"`), so only external-URL anchors carry it; div fallbacks and tel: links no longer get an invalid/needless rel.

File: app/services/whatsapp/baileys/rich_message_parser.rb [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+
+    { type: type, title: title.presence, body: body.presence, footer: footer.presence, buttons: cleaned.presence }.compact
+  end
+
+  def parse_json(str)
+    return {} if str.blank?
+
+    JSON.parse(str)
+  rescue JSON::ParserError
+    {}
```
[the remark this answers is not in the document]
Comment (gabrieljablonski) [reply app/services/whatsapp/baileys/rich_message_parser.rb:200]: Fixed — parse_json now rescues TypeError as well, so a non-string buttonParamsJson (e.g. an already-parsed Hash) degrades to {} instead of raising. Added a regression spec.

**Comment by gabrieljablonski** [comment]
@coderabbitai full review

File: app/javascript/dashboard/components-next/message/bubbles/RichMessage.vue [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    if (button.phone) {
+      return {
+        text: button.text,
+        href: `tel:${button.phone}`,
+        isLink: false,
+        icon: 'i-lucide-phone',
+      };
+    }
+    return { text: button.text, href: null, isLink: false, icon: null };
+  })
```
[the remark this answers is not in the document]
Comment (gabrieljablonski) [reply app/javascript/dashboard/components-next/message/bubbles/RichMessage.vue]: Fixed — the buttons computed now normalizes a non-array payload to [] and filters to plain objects before mapping; tel: links are only built from string phones (toHttpUrl already guards non-strings via try/catch).

File: app/javascript/dashboard/components-next/message/Message.vue [hunk 4]
```diff
@@ -373,6 +374,10 @@ const componentToRender = computed(() => {
     if (fileType === ATTACHMENT_TYPES.CONTACT) return ContactBubble;
   }
 
+  if (props.contentAttributes?.rich) {
+    return RichMessageBubble;
+  }
```
Comment (gabrieljablonski) [inline app/javascript/dashboard/components-next/message/Message.vue:379]: Fixed — shouldRenderMessage now treats a message with contentAttributes.rich as renderable, so the RichMessageBubble branch can't be bypassed.

File: app/services/whatsapp/baileys_handlers/concerns/message_creation_handler.rb [hunk 5]
```diff
@@ -23,6 +26,36 @@ def build_and_save_message(conversation:, sender:, attach_media: false)
...
+    return if fields[:phone].blank? && fields[:name].blank?
+
+    message = conversation.messages.build(
+      content: baileys_contact_line(contact), account_id: inbox.account_id, inbox_id: inbox.id,
+      source_id: raw_message_id, sender: incoming? ? sender : nil,
+      message_type: incoming? ? :incoming : :outgoing, content_attributes: build_message_content_attributes
+    )
+    attach_contact_card(message, fields)
+    message.save!
+    message
```
[the remark this answers is not in the document]
Comment (gabrieljablonski) [reply app/services/whatsapp/baileys_handlers/concerns/message_creation_handler.rb:49]: Declining this one: reusing the parent message id as source_id for every contact row is intentional and mirrors the Cloud provider's create_contact_messages (app/services/whatsapp/incoming_message_base_service.rb:189-196), which passes `source_id: message[:id]` for each contact precisely because contact objects have no id of their own. Diverging here would (a) break the source_id-based dedupe at messages_upsert.rb:35 (`return if find_message_by_source_id(raw_message_id)`), causing duplicate contact rows on every webhook retry, and (b) break in_reply_to/reaction anchoring, which resolves provider ids through source_id (InReplyToMessageBuilder). Keeping it consistent with the established Cloud contract.

File: app/services/whatsapp/baileys/rich_message_parser.rb [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+    PARSERS.lazy.filter_map { |parser| send(parser) }.first
+  end
+
+  # contextInfo of the current rich subtype, where externalAdReply (CTWA) lives.
+  def context_info
+    @msg.dig(:templateMessage, :contextInfo) ||
+      @msg.dig(:interactiveMessage, :contextInfo) ||
+      @msg.dig(:buttonsMessage, :contextInfo) ||
+      @msg.dig(:listMessage, :contextInfo) ||
+      @msg.dig(:interactiveResponseMessage, :contextInfo)
```
[the remark this answers is not in the document]
Comment (gabrieljablonski) [reply app/services/whatsapp/baileys/rich_message_parser.rb:67]: Fixed — context_info now reads via interactive_payload, which covers both interactiveMessage and the nested templateMessage.interactiveMessageTemplate, so externalAdReply/stanzaId from real WABA interactive templates is captured. Added a unit spec.

**Comment by gabrieljablonski** [comment]
@coderabbitai full review

**Comment by gabrieljablonski** [comment]
@coderabbitai full review

File: app/services/whatsapp/baileys/rich_message_parser.rb [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+    { type: type, title: title.presence, body: body.presence, footer: footer.presence, buttons: cleaned.presence }.compact
+  end
+
+  def parse_json(str)
+    return {} if str.blank?
+
+    JSON.parse(str)
+  rescue JSON::ParserError, TypeError
+    {}
+  end
```
[the remark this answers is not in the document]
Comment (gabrieljablonski) [reply app/services/whatsapp/baileys/rich_message_parser.rb:201]: Fixed — parse_json now returns the parsed value only when it is a Hash, otherwise {}. A valid-but-non-object buttonParamsJson (array/scalar) degrades to {} so native_flow_buttons can safely read params['display_text']. Added a spec for the valid-JSON-non-object case.

**Comment by gabrieljablonski** [comment]
@coderabbitai full review

**Comment by gabrieljablonski** [comment]
@coderabbitai full review

File: app/javascript/dashboard/components-next/message/Message.vue (hunk 4 above)
[the remark this answers is not in the document]
Comment (gabrieljablonski) [reply app/javascript/dashboard/components-next/message/Message.vue:379]: Fixed — contextMenuEnabledOptions.delete now also enables on a rich payload (hasText || hasAttachments || hasRichContent), matching the render gate (shouldRenderMessage), so rich-only messages can be deleted. The second instance you referenced (shouldRenderMessage, ~506-516) already had the hasRichContent guard from the previous round.

**Comment by gabrieljablonski** [comment]
@coderabbitai full review
